src/compiler/scala/tools/nsc/backend/jvm/opt/Inliner.scala

```diff
@@ -221,26 +213,82 @@ class Inliner[BT <: BTypes](val btypes: BT) {
     impl(post, mainCallsite)
   }
 
+  class UndoLog(active: Boolean = true) {
+    private var actions = List.empty[() => Unit]
+    private var methodStateSaved = false
+
+    def apply(a: => Unit): Unit = if (active) actions = (() => a) :: actions
+    def run(): Unit = if (active) actions.foreach(_.apply())
+
+    private def arr[T: reflect.ClassTag](l: java.util.List[T]): Array[T] = {
+      val a: Array[T] = new Array[T](l.size)
+      l.toArray(a.asInstanceOf[Array[T with Object]]).asInstanceOf[Array[T]]
```

Comment:
I basically need to save all elements into some new collection, an Array seemed most appropriate because `j.u.List` provides `toArray`. I couldn't get it to compile without casts, not sure if that's a Scala compiler bug. IRC, Scala interprets all java-defined type parameters with `<: Object`.